torch/csrc/jit/codegen/cuda/compute_at_map.cpp

```diff
@@ -12,49 +12,61 @@ namespace fuser {
 namespace cuda {
 namespace {
 
-//! Class to figure out how many non-broadcast axes were used to produce an iter
-//! domain. This is important for figuring out what the correct broadcasted
-//! extent is of an iteration domain.
+//! Class to figure out how many non-broadcast axes and how many broadcast axes
+//! were used to produce an iter domain. This is important for figuring out what
+//! the correct broadcasted extent is of an iteration domain.
 //!
 //! When GpuLower is available, trivial reductions are not counted as
 //! concrete domains so that they should not be used to generate
 //! for-loops.
-class ConcreteInputCounter : public IterVisitor {
+class InputDomainCounter : public IterVisitor {
  public:
-  // Returns number of non-braodcast non-reduction iteration domains used to
-  // generate the iteration domains in provided target domain.
-  static std::unordered_map<IterDomain*, int> produceCounts(
+  // Returns number of {non-braodcast non-reduction iteration domains, broadcast
+  // and trivial reduction domains} used to generate the iteration domains in
+  // provided target domain.
+  static std::unordered_map<IterDomain*, std::pair<int, int>> produceCounts(
       const std::vector<IterDomain*>& domain,
       GpuLower* gpu_lower) {
-    std::unordered_map<IterDomain*, int> count_map;
     if (domain.empty()) {
-      return count_map;
+      return std::unordered_map<IterDomain*, std::pair<int, int>>();
     }
-    ConcreteInputCounter counter(domain, gpu_lower);
-    std::transform(
-        counter.concrete_domain_set_.begin(),
-        counter.concrete_domain_set_.end(),
-        std::inserter(count_map, count_map.begin()),
-        [](const std::pair<IterDomain*, std::unordered_set<IterDomain*>>&
-               entry) {
-          return std::make_pair(entry.first, entry.second.size());
-        });
+
+    InputDomainCounter counter(domain, gpu_lower);
+
+    std::unordered_map<IterDomain*, std::pair<int, int>> count_map;
+    for (auto entry : counter.domain_set_) {
+      auto id = entry.first;
+      auto input_id_set = entry.second;
+      int concrete_counts = 0;
+      int broadcast_counts = 0;
+      for (auto input_id : input_id_set) {
+        if (input_id->isBroadcast() ||
+            (gpu_lower &&
```

Comment:
Can `gpu_lower` ever be null? It seems it should always be non-null. If so, shouldn't we assert it's non-null instead of allowing it to be null?

Reply:
Sorry, I forgot `ComputeAtMap` is also used before lowering, so `gpu_lower` can be null.